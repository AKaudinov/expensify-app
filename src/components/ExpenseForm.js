import React from 'react';
import {connect} from 'react-redux'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component{
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '', //convert the number to a number with cents by dividing by 100
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error:''
        };
    }

    onDescriptionChange = (e) =>{
      const description = e.target.value;
      this.setState(state => {
          return {
              description: description
          }
      });
    };

    onNoteChange = (e) => {
      const note = e.target.value;
      this.setState(prevState => {
         return {
             note: note //don't use e.target.value right in the callback without using e.persist(),
             //since this callback doesn't run right away, and by that time, e.target.value is lost
         }
      });
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //only see numbers that might have a decimal point, and a max of 2 numbers afterwards
            this.setState(prevState => ({
                amount: amount
            }));
        }
    };

    onDateChange = (createdAtMomentDate) => {
        if(createdAtMomentDate) { //if a data is actually passed in, set the state
            this.setState(prevState => {
                return {
                    createdAt: createdAtMomentDate
                }
            })
        }
    };

    onFocusChange = ({focused}) => { //destructure the first argument(focused) that gets passed in as part of an object
        this.setState(prevState => ({
            calendarFocused: focused
        }))
    };

    onSubmitForm = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            //set error state equal to 'Please provide description and amount'
            this.setState(prevState => {
                return{
                    error: 'Please provide description and amount'
                }
            })
        }else{
            //clear the error
            this.setState(prevState => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //format the number correctly
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render(){
        return (
            <div>
                {this.state.error && <h4>{this.state.error}</h4>}
              <form onSubmit={this.onSubmitForm}>
                  <input
                  type="text"
                  placeholder="Description"
                  autoFocus
                  value={this.state.description}
                  onChange={this.onDescriptionChange}/>

                  <input
                  type="text"
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}/>

                  <SingleDatePicker
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  numberOfMonths={1}
                  isOutsideRange={day => false}/>

                  <textarea
                      placeholder="Add a note for your expense (optional)"
                      value={this.state.note}
                      onChange={this.onNoteChange}>
                  </textarea>

                  <button>Save expense</button>

              </form>
            </div>
        )
    }
}
