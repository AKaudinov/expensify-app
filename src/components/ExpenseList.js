import React from 'react';
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) =>{
    return (
        <div>
            <h1>Expense List: </h1>
            {props.expenses.map(expense =>
               (<ExpenseListItem
                expense={expense}
                key={expense.id}/>))}
        </div>
    );
};

const mapStateToProps = (state) =>{
    return {
        expenses: selectExpenses(state.expenses, state.filters), //grabs the information from the store, information that our reducers return, in this case, expenses reducer
        filters: state.filters //grabs filters from the filters reducer
    }
};

// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses //grabs the information from the store, information that our reducers return, in this case, expenses reducer
//     }
// })(ExpenseList);

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList); //connect() returns a function that you need to call
export default ConnectedExpenseList;

// OR:


// export default connect(mapStateToProps)(ExpenseList); //connect() returns a function that you need to call
