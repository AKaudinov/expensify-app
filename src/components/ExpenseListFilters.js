import React from 'react';
import {connect} from 'react-redux';
import * as filterActions from '../actions/filters';
import {DateRangePicker} from 'react-dates';


export class ExpenseListFilters extends React.Component{
    state={
      calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        // this.props.dispatch(filterActions.setStartDate(startDate));
        // this.props.dispatch(filterActions.setEndDate(endDate));
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = calendarFocused => {
        this.setState(prevState => ({
            calendarFocused: calendarFocused
        }));
    };

    onTextChange = (e) =>{
        // this.props.dispatch(filterActions.setTextFilter(e.target.value))
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = e => {
        if(e.target.value === 'date'){
            // this.props.dispatch(filterActions.sortByDate())
            this.props.sortByDate();
        }else if(e.target.value === 'amount'){
            // this.props.dispatch(filterActions.sortByAmount())
            this.props.sortByAmount();
        }
    };

    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            placeholder="Search expenses"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}/>
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy} //controlled input - value controlled by JavaScript
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={day => false}/>
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
  return{
    filters: state.filters
  };
};

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(filterActions.setTextFilter(text)),
    sortByDate: () => dispatch(filterActions.sortByDate()),
    sortByAmount: () => dispatch(filterActions.sortByAmount()),
    setStartDate: startDate => dispatch(filterActions.setStartDate(startDate)),
    setEndDate: endDate => dispatch(filterActions.setEndDate(endDate))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
