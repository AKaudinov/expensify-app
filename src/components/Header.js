import React from 'react'
import {NavLink} from "react-router-dom";

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" exact={true} activeClassName="active-link">Home</NavLink>
        <NavLink to="/create" activeClassName="active-link">Create Expense</NavLink>
        {/*<NavLink to="/edit" activeClassName="active-link">Edit Expense</NavLink>*/}
        {/*<NavLink to="/help" activeClassName="active-link">Help</NavLink>*/}
    </header>
);

export default Header;
