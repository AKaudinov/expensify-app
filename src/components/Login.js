import React from 'react';
import {connect} from 'react-redux';
import * as loginActions from '../actions/auth';

export class Login extends React.Component{
    constructor(props) {
        super(props);
    }

    onLogin = (e) =>{
        e.preventDefault(); //I'm an idiot because I didn't prevent default here and the page was refreshing before firebase had a chance to log in!
        this.props.login();
    };

    render(){
        return (
            <div>
            <form onSubmit={this.onLogin}>
                <input type="text"/>
                <input type="text"/>
                <button>Login</button>
            </form>
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        login: () => dispatch(loginActions.startLogin())
    }
};

export default connect(undefined, mapDispatchToProps)(Login);

