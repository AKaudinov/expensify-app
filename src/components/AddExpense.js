import React from 'react';
import ExpenseForm from "./ExpenseForm";
import {connect} from 'react-redux';
import * as expenseActions from '../actions/expenses';

export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

// const AddExpense = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//         onSubmit={(expenseObj) => {
//             //props.dispatch(expenseActions.addExpense(expenseObj));
//             props.addExpense(expenseObj);
//             props.history.push('/'); //redirect back to the dashboard page, which in this case is home page
//         }}/>
//     </div>
// );

//similar to state to props
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(expenseActions.addExpense(expense))
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpense);
