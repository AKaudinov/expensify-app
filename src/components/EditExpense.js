import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import * as expenseActions from '../actions/expenses';

export class EditExpense extends React.Component{
    onSubmit = (expense) => {
        this.props.editExpense(this.props.match.params.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
      this.props.removeExpense(this.props.expense);
      this.props.history.push('/');
    };

    render(){
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}/>
                <button onClick={this.onRemove}>Remove Expense</button>
            </div>
        );
    }
}

// const EditExpense = (props) => {
//     return (
//     <div>
//         <ExpenseForm
//             expense={props.expense}
//         onSubmit={expenseObj => {
//             props.dispatch(expenseActions.editExpense(props.match.params.id, expenseObj));
//             props.history.push('/');
//         }}/>
//         <button onClick={() => {
//             props.dispatch(expenseActions.removeExpense(props.expense));
//             props.history.push('/');
//         }}>Remove Expense</button>
//     </div>
//     );
// };

const mapStateToProps = (state, props) => { //find the expense with the given id from the passed in url
    return {
      expense: state.expenses.find(expenseItem => expenseItem.id === props.match.params.id) //search through the array and find the single item)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
      editExpense: (id, expense) => dispatch(expenseActions.editExpense(id, expense)),
        removeExpense: (expense) => dispatch(expenseActions.removeExpense(expense))
    }
};

const connectedEditExpense = connect(mapStateToProps, mapDispatchToProps)(EditExpense);
export default connectedEditExpense;
