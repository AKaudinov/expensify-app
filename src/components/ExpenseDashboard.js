import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboard = (props) => (
    <div>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);


export default ExpenseDashboard;
