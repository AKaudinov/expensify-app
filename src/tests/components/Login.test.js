import React from 'react';
import {shallow} from 'enzyme';
import {Login} from "../../components/Login";
import {Header} from "../../components/Header";

test('should render login component', () => {
   const wrapper = shallow(<Login/>);
   expect(wrapper).toMatchSnapshot();
});


test('Should call login on submit click', () => {
   const login = jest.fn();
   const wrapper = shallow(<Login login={login}/>);

   wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
   });
   expect(login).toHaveBeenCalled();

});

// test('Should call logout on button click', () => {
//    const logout = jest.fn();
//    const wrapper = shallow(<Header logout={logout}/>);
//    wrapper.find('button').simulate('click');
//
//    expect(logout).toHaveBeenCalled();
// });
