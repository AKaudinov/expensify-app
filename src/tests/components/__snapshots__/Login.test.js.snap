// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login component 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <button>
      Login
    </button>
  </form>
</div>
`;
